src/nnbench/types.py

```diff
@@ -110,22 +117,72 @@ class LocalArtifactLoader(ArtifactLoader):
     Parameters
     ----------
     path : str | os.PathLike[str]
-        The file system pathto the artifact.
+        The file system path to the artifact.
     """
 
-    def __init__(self, path: str | os.PathLike[str]):
+    def __init__(self, path: str | os.PathLike[str]) -> None:
         self._path = path
 
-    def load(self):
+    def load(self) -> Path:
         """
         Returns the path to the artifact on the local file system.
         """
-        return self._path
+        return Path(self._path).resolve()
 
 
-class S3ArtifactLoader(ArtifactLoader):
-    # TODO: Implement this and other common ArtifactLoders here or in a util
-    pass
+class FilePathArtifactLoader(ArtifactLoader):
+    """
+    ArtifactLoader for loading artifacts using fsspec-supported file systems.
+
+    This allows for loading from various file systems like local, S3, GCS, etc.,
+    by using a unified API provided by fsspec.
+
+    Parameters
+    ----------
+    path : str | os.PathLike[str]
+        The path to the artifact, which can include a protocol specifier (like 's3://') for remote access.
+    to_localdir : str | os.PathLike[str] | None
+        The local directory to which remote artifacts will be downloaded. If provided, the model data will be persisted. Otherwise, local artifacts are cleaned.
+    storage_options : dict[str, Any] | None
+        Storage options for remote storage.
+    """
+
+    def __init__(
+        self,
+        path: str | os.PathLike[str],
+        to_localdir: str | os.PathLike[str] | None = None,
+        storage_options: dict[str, Any] | None = None,
+    ) -> None:
+        self.source_path = str(path)
+        if to_localdir:
+            self.target_path = Path(to_localdir)
+            delete = False
+        else:
+            self.target_path = Path(mkdtemp())
+            delete = True
+        self._finalizer = weakref.finalize(self, self._cleanup, delete=delete)
+        self.storage_options = storage_options or {}
+
+    def load(self) -> Path:
+        """
+        Loads the artifact, downloading it if necessary, and returns the local path.
+
+        Returns
+        -------
+        Path
+            The path to the artifact on the local filesystem.
+        """
+        fs, _, _ = core.get_fs_token_paths(self.source_path)
+        if isinstance(fs, LocalFileSystem):
+            return Path(self.source_path).resolve()
```

Comment:
Is this caching a la "check if it exists, only download if it doesn't"? (Otherwise it's probably currently missing)

Reply:
More or less.

My understanding is, that this class, as a general filesystem class, should also be able to use local filepaths.
In that case, we don't want to download the data.

If it is a remote path, we download and put it either (a) a specified local dir,  or (b) a tmp dir. 

If a) the files are persisted on disk and the filepath can be used (by the user) in other places. 

If b) the cleanup task deletes the tmp dir and the model to not clutter the filesystem. 

That is how it is currently implemented. Do you agree with this?